test/unit/tools/test_metadata.py

```diff
@@ -0,0 +1,171 @@
+from galaxy.metadata import get_metadata_compute_strategy
+from galaxy.objectstore import ObjectStorePopulator
+from .. import tools_support
+
+
+class MetadataTestCase(unittest.TestCase, tools_support.UsesApp, tools_support.UsesTools):
+
+    def setUp(self):
+        super(MetadataTestCase, self).setUp()
+        self.setup_app()
+        model.Dataset.object_store = self.app.object_store
+        job = model.Job()
+        sa_session = self.app.model.session
+        sa_session.add(job)
+        history = model.History()
+        job.history = history
+        sa_session.flush()
+        self.job = job
+        self.history = history
+        self.job_working_directory = os.path.join(self.test_directory, "job_working")
+        self.tool_working_directory = os.path.join(self.job_working_directory, "working")
+        os.mkdir(self.job_working_directory)
+        os.mkdir(self.tool_working_directory)
+
+    def tearDown(self):
+        super(MetadataTestCase, self).tearDown()
+        self.metadata_compute_strategy = None
+
+    def test_simple_output(self):
+        source_file_name = os.path.join(os.getcwd(), "test/functional/tools/for_workflows/cat.xml")
+        self._init_tool_for_path(source_file_name)
+        output_dataset = self._create_output_dataset(
+            extension="fasta",
+        )
+        sa_session = self.app.model.session
+        sa_session.flush()
+        output_datasets = {
+            "out_file1": output_dataset,
+        }
+        command = self.metadata_command(output_datasets)
+        # self._write_galaxy_json("""{"type": "dataset", "dataset_id": "%s", "name": "my dynamic name", "ext": "fasta", "info": "my dynamic info"}""" % output_dataset.id)
+        self._write_output_dataset_contents(output_dataset, ">seq1\nGCTGCATG\n")
+        self.exec_metadata_command(command)
+        metadata_set_successfully = self.metadata_compute_strategy.external_metadata_set_successfully(output_dataset, sa_session)
+        assert metadata_set_successfully
+        self.metadata_compute_strategy.load_metadata(output_dataset, "out_file1", sa_session, working_directory=self.job_working_directory)
+        assert output_dataset.metadata.data_lines == 2
+        assert output_dataset.metadata.sequences == 1
+
+    def test_primary_dataset_output_extension(self):
+        source_file_name = os.path.join(os.getcwd(), "test/functional/tools/for_workflows/cat.xml")
+        self._init_tool_for_path(source_file_name)
+        output_dataset = self._create_output_dataset(
+            extension="auto",
+        )
+        sa_session = self.app.model.session
+        sa_session.flush()
+        output_datasets = {
+            "out_file1": output_dataset,
+        }
+        command = self.metadata_command(output_datasets)
+        self._write_galaxy_json("""{"type": "dataset", "dataset_id": "%s", "name": "my dynamic name", "ext": "fasta", "info": "my dynamic info"}""" % output_dataset.dataset.id)
```

Comment:
should that be `auto`?

Reply:
No - setting it to auto during setup results in the extension being set in this file being respected. So it should be auto above and actual extension here. I'll leave a comment to this point - because I agree it reads funny.